compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -661,4 +683,74 @@ impl<'context> Elaborator<'context> {
             Err(error) => make_error(self, error),
         }
     }
+
+    fn try_get_comptime_function(
+        &mut self,
+        func: ExprId,
+        location: Location,
+    ) -> Result<FuncId, ResolverError> {
+        match self.interner.expression(&func) {
+            HirExpression::Ident(ident, _generics) => {
+                let definition = self.interner.definition(ident.id);
+                if let DefinitionKind::Function(function) = definition.kind {
+                    let meta = self.interner.function_modifiers(&function);
+                    if meta.is_comptime {
+                        Ok(function)
+                    } else {
+                        Err(ResolverError::MacroIsNotComptime { span: location.span })
+                    }
+                } else {
+                    Err(ResolverError::InvalidSyntaxInMacroCall { span: location.span })
+                }
+            }
+            _ => Err(ResolverError::InvalidSyntaxInMacroCall { span: location.span }),
+        }
+    }
+
+    /// Call a macro function and inlines its code at the call site.
+    /// This will also perform a type check to ensure that the return type is an `Expr` value.
+    fn call_macro(
+        &mut self,
+        func: ExprId,
+        arguments: Vec<ExprId>,
+        location: Location,
+        return_type: Type,
+    ) -> Option<(HirExpression, Type)> {
+        self.unify(&return_type, &Type::Expr, || TypeCheckError::MacroReturningNonExpr {
+            typ: return_type.clone(),
+            span: location.span,
+        });
+
+        let function = match self.try_get_comptime_function(func, location) {
+            Ok(function) => function,
+            Err(error) => {
+                self.push_err(error);
+                return None;
+            }
+        };
+
+        let mut interpreter = Interpreter::new(self.interner, &mut self.comptime_scopes);
+
+        let mut comptime_args = Vec::new();
+        let mut errors = Vec::new();
+
+        for argument in arguments {
+            match interpreter.evaluate(argument) {
+                Ok(arg) => {
+                    let location = interpreter.interner.expr_location(&argument);
+                    comptime_args.push((arg, location));
+                }
+                Err(error) => errors.push((error.into(), self.file)),
+            }
+        }
+
+        if !errors.is_empty() {
```

Comment:
Did you do this so that we can collect all errors for each comptime arg in one run? We could avoid this if statement by push to `self.errors` and returning `None` in the match above.

Reply:
Yep, I wanted to evaluate all the arguments since they're mostly independent.